src/bartiq/_routine.py

```diff
@@ -66,380 +45,190 @@ class PortDirection(str, Enum):
-    meta: Optional[dict[str, Any]] = Field(default_factory=dict)
-
-    def __repr__(self):
-        parent_name = "none" if self.parent is None else self.parent.name
-        size_value = "None" if self.size is None else f'"{self.size}"'
-        return f"{self.__class__.__name__}({parent_name}.#{self.name}, size={size_value}, {self.direction})"
-
-    def absolute_path(self, exclude_root_name: bool = False) -> str:
-        """Returns a path from root.
-
-        Args:
-            exclude_root_name: If true, excludes name of root from the path. Default: False
-        """
-        assert self.parent is not None
-        if self.parent.absolute_path(exclude_root_name=exclude_root_name) == "":
-            return f"#{self.name}"
-        else:
-            return f"{self.parent.absolute_path(exclude_root_name=exclude_root_name)}.#{self.name}"
-
-
-class Connection(BaseModel):
-    """Connection between two ports.
-
-    Attributes:
-        source: Port which the connection comes from.
-        target: Port the connection targets.
-        parent: Routine this connection belongs to. Note: it is marked as Optional
-            only because of how Routine objects are internally constructed. In
-            correctly constructed routines, no actual connection should have
-            a None as a parent.
-    """
-
-    source: Port
-    target: Port
-    parent: Optional[Routine] = Field(exclude=True, default=None)
-
-    def __repr__(self):
-        parent_name = "none" if self.parent is None else self.parent.name
-        return f"{self.__class__.__name__}({parent_name}.#{self.source.name} -> {parent_name}.#{self.target.name})"
-
-    @field_serializer("source", "target")
-    def _serialize_port(self, port):
-        return port.name if port.parent is self.parent else f"{port.parent.name}.{port.name}"
+    def from_qref(cls, qref_obj: SchemaV1 | RoutineV1, backend: SymbolicBackend[T]) -> CompiledRoutine[T]:
+        program = qref_obj.program if isinstance(qref_obj, SchemaV1) else qref_obj
+        return CompiledRoutine[T](
+            children={child.name: cls.from_qref(child, backend) for child in program.children},
+            **_common_routine_dict_from_qref(qref_obj, backend),
+        )
+
+
+def _common_routine_dict_from_qref(
+    qref_obj: SchemaV1 | RoutineV1, backend: SymbolicBackend[T]
+) -> _CommonRoutineParams[T]:
+    program = qref_obj.program if isinstance(qref_obj, SchemaV1) else qref_obj
+    return {
+        "name": program.name,
+        "type": program.type,
+        "ports": {port.name: _port_from_qref(port, backend) for port in program.ports},
+        "input_params": tuple(program.input_params),
+        "resources": {resource.name: _resource_from_qref(resource, backend) for resource in program.resources},
+        "connections": {
+            _endpoint_from_qref(conn.source): _endpoint_from_qref(conn.target) for conn in program.connections
+        },
+    }
+
+
+def _rsplit_once(target: str) -> tuple[str, str]:
```

Comment:
Originally this was to satisfy the typechecker (which actually was right). We use output from this function to create tuples of the form (str, str). To see why the typechecker complained, consider this dummy example:

```python
# dummy function accepting pairs of strings
def foo(s: tuple[str, str]) -> None:
    print(s[0], s[1])

# example usage, which is wrong
s = input()
foo(tuple(s.rsplit(".", 1)))
```

The usage shown above is wrong, because `rsplit(".", 1)` can produce a list of length 1. I circumvented it by making a function `_rsplit_once` that ensures that we get two strings from `rsplit`.

Having said that, I just came up with a simpler solution, see [3351962](https://github.com/PsiQ/bartiq/pull/115/commits/3351962fdf453f2b47acc681c861c7013519a4fa).